cdap-securestore-ext-cloudkms/src/main/java/co/cask/cdap/securestore/spi/CloudKMSClient.java

```diff
@@ -0,0 +1,226 @@
+import com.google.api.client.http.javanet.NetHttpTransport;
+import com.google.api.client.json.JsonFactory;
+import com.google.api.client.json.jackson2.JacksonFactory;
+import com.google.api.services.cloudkms.v1.CloudKMS;
+import com.google.api.services.cloudkms.v1.CloudKMSScopes;
+import com.google.api.services.cloudkms.v1.model.CryptoKey;
+import com.google.api.services.cloudkms.v1.model.DecryptRequest;
+import com.google.api.services.cloudkms.v1.model.DecryptResponse;
+import com.google.api.services.cloudkms.v1.model.EncryptRequest;
+import com.google.api.services.cloudkms.v1.model.EncryptResponse;
+import com.google.api.services.cloudkms.v1.model.KeyRing;
+import com.google.common.io.CharStreams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashSet;
+import java.util.Set;
+import javax.annotation.Nullable;
+
+/**
+ * Wrapper on {@link CloudKMS} client.
+ */
+class CloudKMSClient implements Closeable {
+  private static final Logger LOG = LoggerFactory.getLogger(CloudKMSClient.class);
+  // When created in the global location, Cloud KMS resources are available from zones spread around the world.
+  private static final String LOCATION_ID = "global";
+  // All the crypto keys are created under keyring named cdap
+  private static final String KEYRING_ID = "cdap";
+  private static final String ENCRYPT_DECRYPT = "ENCRYPT_DECRYPT";
+  private static final String CLOUD_KMS = "cloudkms";
+
+  private final CloudKMS cloudKMS;
+  private final String projectId;
+  // In-memory cache to hold created crypto keys, this is to avoid checking if a given crypto key exists.
+  private final Set<String> crypoKeyList;
+
+  /**
+   * Constructs Cloud KMS client.
+   *
+   * @throws IOException if cloud kms client can not be created
+   */
+  CloudKMSClient() throws IOException {
+    this.crypoKeyList = new HashSet<>();
+    this.projectId = getSystemProjectId();
+    this.cloudKMS = createAuthorizedClient();
+  }
+
+  /**
+   * Get project id from the metadata server. Makes a request to the metadata server that lives on the VM,
+   * as described at https://cloud.google.com/compute/docs/storing-retrieving-metadata.
+   */
+  private static String getSystemProjectId() {
+    try {
+      URL url = new URL("http://metadata.google.internal/computeMetadata/v1/project/project-id");
+      HttpURLConnection connection = null;
+      try {
+        connection = (HttpURLConnection) url.openConnection();
+        connection.setRequestProperty("Metadata-Flavor", "Google");
+        connection.connect();
+        try (Reader reader = new InputStreamReader(connection.getInputStream(), StandardCharsets.UTF_8)) {
+          return CharStreams.toString(reader);
+        }
+      } finally {
+        if (connection != null) {
```

Comment:
Move the assignment to `connection` outside of the `try`, then you don't need to check for `null` here.

Reply:
Also, how to set projectId explicitly? This method is called from the default constructor.